### PR TITLE
Improve exceptions

### PR DESCRIPTION
**Pull request by Art4**
At the moment I have the problem that exceptions thrown by the library cannot be distinguished from other exceptions thrown for other reasons. This makes testing with the library a bit harder.

At the moment we throw only two different exceptions:

1. `\Exception`, when something fails during the transfer to Redmine or if an API parameter is invalid or missing
2. `\InvalidArgumentException` when a non-existent API is requested from the client.

This PR introduces these new exceptions that extend the previous exceptions:

1. `\Redmine\Exception\ClientException` replaces `\Exception` in CurlClient and Psr18Client.
2. `\Redmine\Exception\InvalidApiNameException` replaces `\InvalidArgumentException('... is not a valid api. Possible apis are ...')`.
3. `\Redmine\Exception\MissingParameterException` replaces `\Exception('Missing mandatory parameters')`.
4. `\Redmine\Exception\InvalidParameterException` replaces `\Exception('Possible values for ...')`.

Additionally these new exceptions implement the interface `\Redmine\Exception`. 

This makes it easier to catch only the exceptions thrown by the Redmine library. Nevertheless, we preserve backward compatibility.

```php
try {
    $client->getApi('issues')->create($data);
} catch (\Redmine\Exception $e) {
    // exceptions from kbsali/redmine-api
} catch (\Throwable $e) {
    // other exceptions
}
```

What do you think, @kbsali? If you like the idea, I will create new tests for the new exceptions.

### CODE REVIEW

File: src/Redmine/Exception.php [hunk 1]
```diff
@@ -0,0 +1,12 @@
+<?php
+
+namespace Redmine;
+
+use Throwable;
+
+/**
+ * Exception interface.
```
Comment (kbsali) [inline src/Redmine/Exception.php]: redundant! :) (same for the other exceptions)

**Review by kbsali — COMMENTED** [review]
I love the idea! :+1:

**Comment by Art4** [comment]
> I love the idea! 👍

Thank you. I will finalize the PR.

Btw what do you think about replacing travic-ci with Github Actions?

**Comment by Art4** [comment]
I've found two more cases where exceptions can be thrown and added two new Exception: `MissingParameterException` and `InvalidParameterException`.

**Review by kbsali — APPROVED** [review]
:+1:

**Comment by Art4** [comment]
This PR is ready for review. Please take a special look at the wording in the docs. 😊

File: src/Redmine/Api/Group.php [hunk 2]
```diff
@@ -56,7 +59,7 @@ public function listing($forceUpdate = false)
      *
      * @param array $params the new group data
      *
-     * @throws \Exception Missing mandatory parameters
+     * @throws MissingParameterException Missing mandatory parameters
```
Comment (kbsali) [inline src/Redmine/Api/Group.php:62]: now that the exceptions have meaningfull names, maybe those annotations are redundant, don't you think?

File: docs/usage.md [hunk 3]
```diff
@@ -195,6 +199,22 @@ $client->startImpersonateUser('kim');
 $client->stopImpersonateUser();
 ```
 
+## Error handling
+
+Every thrown exception will implement the interface `Redmine\Exception`. This way you can catch all Redmine specific exceptions.
```
Comment (kbsali) [inline docs/usage.md]:
```suggestion
Every exception implement the interface `Redmine\Exception` making it easy to catch Redmine specific issues.
```

File: src/Redmine/Api/Group.php [hunk 4]
```diff
@@ -71,7 +74,7 @@ public function create(array $params = [])
         if (
             !isset($params['name'])
         ) {
-            throw new \Exception('Missing mandatory parameters');
+            throw new MissingParameterException('Theses parameters are mandatory: `name`');
```
Comment (kbsali) [inline src/Redmine/Api/Group.php:77]:
```suggestion
            throw new MissingParameterException('Missing mandatory parameters `name`');
```
?

**Review by kbsali — APPROVED** [review]
very nice! Thanks for this big PR and sorry for the late review.
Only minor comments, you could decide not to address them! :)

File: src/Redmine/Api/Group.php (hunk 2 above)
Comment (Art4) [reply src/Redmine/Api/Group.php:62]: We should keep the `@throws` annotations for better IDE integration.

File: src/Redmine/Api/Group.php (hunk 4 above)
Comment (Art4) [reply src/Redmine/Api/Group.php:77]: A `MissingParameterException` is thrown if one or more mandatory parameters are missing. I tried to find a generic message, that fits in all places but without the need of logic code to name exactly the missing parameter/paremeters.

**Pull request merged**